File: src/components/table/basic/tableCellBadge.test.js
import assert from 'node:assert/strict';

import {resolveTableCellBadge} from './tableCellBadge.js';

const context = {signals: {authorization: {peek: () => ({principal: {features: ['FEATURE_PRIORITY']}})}}};
const badge = {
  label: 'Priority',
  icon: 'endorsed',
  field: 'planType',
  equals: 2,
  tone: 'accent',
  visibleWhen: {source: 'authorization', field: 'principal.features', contains: 'FEATURE_PRIORITY'},
};

assert.equal(resolveTableCellBadge({planType: 1}, badge, context), null);
assert.deepEqual(resolveTableCellBadge({planType: 2}, badge, context), {
  label: 'Priority', icon: 'endorsed', tone: 'accent', className: '', tooltip: 'Priority', replaceValue: false, hideLabel: false,
});
assert.equal(resolveTableCellBadge({planType: 2}, badge, {signals: {authorization: {peek: () => ({principal: {features: []}})}}}), null);

assert.deepEqual(resolveTableCellBadge({status: 4}, {
  field: 'status',
  valueMap: {'0': 'Inactive', '4': 'Completed'},
  toneMap: {'0': 'neutral', '4': 'info'},
  replaceValue: true,
}), {
  label: 'Completed', icon: '', tone: 'info', className: '', tooltip: 'Completed', replaceValue: true, hideLabel: false,
});

assert.deepEqual(resolveTableCellBadge({status: 0}, {
  field: 'status',
  valueMap: {'0': 'Inactive'},
  toneMap: {'0': 'neutral'},
  replaceValue: true,
}), {
  label: 'Inactive', icon: '', tone: 'neutral', className: '', tooltip: 'Inactive', replaceValue: true, hideLabel: false,
});

assert.deepEqual(resolveTableCellBadge({validationAlert: 'Creative has validation errors'}, {
  field: 'validationAlert', icon: 'warning-sign', tone: 'danger', hideLabel: true,
}), {
  label: '', icon: 'warning-sign', tone: 'danger', className: '', tooltip: 'Creative has validation errors', replaceValue: false, hideLabel: true,
});

console.log('tableCellBadge ✓ conditionally resolves a generic row decoration');

File: backend/types/table_cell_badge_test.go
package types

import (
	"testing"

	"gopkg.in/yaml.v3"
)

func TestColumnRetainsConditionalBadge(t *testing.T) {
	source := []byte(`
id: root
table:
  columns:
    - id: name
      name: Name
      sticky: left
      badge:
        label: Priority
        icon: endorsed
        field: planType
        equals: 2
        tone: accent
        visibleWhen: {source: authorization, field: principal.features, contains: FEATURE_PRIORITY}
`)
	container := &Container{}
	if err := yaml.Unmarshal(source, container); err != nil {
		t.Fatal(err)
	}
	badge := container.Table.Columns[0].Badge
	if container.Table.Columns[0].Sticky != "left" {
		t.Fatalf("sticky column position was not retained: %#v", container.Table.Columns[0])
	}
	if badge == nil || badge.Label != "Priority" || badge.Field != "planType" || badge.Equals != 2 || badge.Tone != "accent" {
		t.Fatalf("conditional badge was not retained: %#v", badge)
	}
	if badge.VisibleWhen["source"] != "authorization" {
		t.Fatalf("badge visibility rule was not retained: %#v", badge.VisibleWhen)
	}
}
